In [1]:
import matplotlib.pyplot as plt
import seaborn as sns

# from your hashing notes
def shingle(text: str, k: int):
    shingle_set = []

    for i in range(len(text) - k+1):
        shingle_set.append(text[i:i+k])
    return set(shingle_set)

#help from ChatGPT for this function
def shingleWords(text: str):
    shingle_set = []
    words = text.split()

    for i in range(len(words) - 1):
        shingle_set.append((words[i], words[i+1]))
    return set(shingle_set)

# from your hashing notes
def jaccard(a: set, b: set):
    return len(a.intersection(b)) / len(a.union(b))

# Character Based 3-grams

In [29]:
with open('Random spanish sentences.txt') as f:
    sentences = [sentences.strip() for sentences in f.readlines()]

#to get rid of repeat sentences
uniqueSentences = set(sentences)
sentences = list(uniqueSentences)

k = 3

#stores all created k-grams in a list
kGrams = []
#stores lists of k-grams created from each sentence
threeGram= []
for x in sentences:
     threeGram.append(shingle(x.lower(), k))
     kGrams += shingle(x.lower(), k)
     
#removes duplicate k-grams
uniqueThreeKGrams = set(kGrams)
print(len(uniqueThreeKGrams))

similarityList = []
#loops through all sentence k-grams and runs the jaccard similarity on all pairs
for i in range(len(threeGram)):
    sentence1 = threeGram[i]
    for j in range(i+1, len(threeGram)):
        similarityList.append(jaccard(sentence1, threeGram[j]))

resultList = []       
for i in range(len(threeGram)):
    for j in range(i+1, len(threeGram)):
        score = jaccard(threeGram[i], threeGram[j])
        resultList.append((sentences[i], sentences[j], score))

for i in range(len(resultList)):
    for j in range(i+1, len(resultList)):
        if resultList[i][2] < resultList[j][2]:
            temp = resultList[i]
            resultList[i] = resultList[j]
            resultList[j] = temp

print(resultList[:5])

1574
[('La tecnología también ha ayudado a reducir los costos de producción al permitir que las empresas realicen muchas tareas de forma más eficiente.', 'La tecnología también ha mejorado la seguridad al permitir que las empresas rastreen y controlen sus datos de forma más eficiente.', 0.47023809523809523), ('El viento soplaba suavemente mientras él cerraba los ojos para disfrutar de la tranquilidad.', 'La brisa suave del mar acariciaba su rostro mientras él cerraba los ojos para disfrutar del momento.', 0.452991452991453), ('El que no se arriesga, no pasa la mar.', 'El que no arriesga, no gana.', 0.45), ('La tecnología ha cambiado la forma en que nos conectamos con los demás al ofrecer nuevas formas de comunicación, como el chat en vivo.', 'La tecnología ha cambiado la forma en que interactuamos con el mundo y ha proporcionado nuevas formas de hacer negocios.', 0.40384615384615385), ('Todos los presentes se unieron a la fiesta, cantando, bailando y disfrutando de la compañía de sus v

# Character Based 2-grams

In [28]:
with open('Random spanish sentences.txt') as f:
    sentences = [sentences.strip() for sentences in f.readlines()]

#to get rid of repeat sentences
uniqueSentences = set(sentences)
sentences = list(uniqueSentences)

k = 2

#stores all created k-grams in a list
kGrams = []
#stores lists of k-grams created from each sentence
twoGram= []
for x in sentences:
     twoGram.append(shingle(x.lower(), k))
     kGrams += shingle(x.lower(), k)

#removes duplicate k-grams
uniqueTwoKGrams = set(kGrams)
print(len(uniqueTwoKGrams))

similarityList = []
#loops through all sentence k-grams and runs the jaccard similarity on all pairs
for i in range(len(twoGram)):
    sentence1 = twoGram[i]
    for j in range(i+1, len(twoGram)):
        similarityList.append(jaccard(sentence1, twoGram[j]))

resultList = []       
for i in range(len(twoGram)):
    for j in range(i+1, len(twoGram)):
        score = jaccard(twoGram[i], twoGram[j])
        resultList.append((sentences[i], sentences[j], score))

for i in range(len(resultList)):
    for j in range(i+1, len(resultList)):
        if resultList[i][2] < resultList[j][2]:
            temp = resultList[i]
            resultList[i] = resultList[j]
            resultList[j] = temp

print(resultList[:5])

371
[('La tecnología también ha ayudado a reducir los costos de producción al permitir que las empresas realicen muchas tareas de forma más eficiente.', 'La tecnología también ha mejorado la seguridad al permitir que las empresas rastreen y controlen sus datos de forma más eficiente.', 0.6696428571428571), ('El viento soplaba suavemente mientras él cerraba los ojos para disfrutar de la tranquilidad.', 'La brisa suave del mar acariciaba su rostro mientras él cerraba los ojos para disfrutar del momento.', 0.5764705882352941), ('La tecnología también ha permitido a muchas empresas llevar a cabo transacciones de forma más segura con el uso de la criptografía.', 'La tecnología también ha mejorado la seguridad al permitir que las empresas rastreen y controlen sus datos de forma más eficiente.', 0.5655737704918032), ('La tecnología ha cambiado la forma en que nos conectamos con los demás al ofrecer nuevas formas de comunicación, como el chat en vivo.', 'La tecnología ha cambiado la forma en q

# Word Based 2-grams

In [30]:
with open('Random spanish sentences.txt') as f:
    sentences = [sentences.strip() for sentences in f.readlines()]

#to get rid of repeat sentences
uniqueSentences = set(sentences)
sentences = list(uniqueSentences)

k = 2

#stores all created k-grams in a list
kGrams = []
#stores lists of k-grams created from each sentence
wordGram= []
for x in sentences:
     wordGram.append(shingleWords(x.lower()))
     kGrams += shingleWords(x.lower())

#removes duplicate k-grams
uniqueWordKGrams = set(kGrams)
print(len(uniqueWordKGrams))


resultList = []       
for i in range(len(wordGram)):
    for j in range(i+1, len(wordGram)):
        score = jaccard(wordGram[i], wordGram[j])
        resultList.append((sentences[i], sentences[j], score))

for i in range(len(resultList)):
    for j in range(i+1, len(resultList)):
        if resultList[i][2] < resultList[j][2]:
            temp = resultList[i]
            resultList[i] = resultList[j]
            resultList[j] = temp

print(resultList[:5])

1245
[('La tecnología también ha ayudado a reducir los costos de producción al permitir que las empresas realicen muchas tareas de forma más eficiente.', 'La tecnología también ha mejorado la seguridad al permitir que las empresas rastreen y controlen sus datos de forma más eficiente.', 0.3125), ('El que no se arriesga, no pasa la mar.', 'El que no arriesga, no gana.', 0.3), ('La tecnología ha cambiado la forma en que nos conectamos con los demás al ofrecer nuevas formas de comunicación, como el chat en vivo.', 'La tecnología ha cambiado la forma en que interactuamos con el mundo y ha proporcionado nuevas formas de hacer negocios.', 0.2727272727272727), ('El viento soplaba suavemente mientras él cerraba los ojos para disfrutar de la tranquilidad.', 'La brisa suave del mar acariciaba su rostro mientras él cerraba los ojos para disfrutar del momento.', 0.2608695652173913), ('El aire era fresco y limpio y la brisa suave del mar acariciaba su cara.', 'La brisa suave del mar acariciaba su r

## Distinct k-grams
#### Very Short Sentences
2-gram: 162 <br>
3-gram: 277 <br>
word 2-gram: 89<br>
#### Short Sentences
2-gram: 288 <br>
3-gram: 821 <br>
word 2-gram: 367 <br>
#### Long Sentences
2-gram: 395 <br>
3-gram: 1804 <br>
word 2-gram: 1674 <br>
#### Spanish Sentences
2-gram: 371 <br>
3-gram: 1574 <br>
word 2-gram: 1245 <br>

## 5 Pairs of Sentences
#### Very Short Sentences 
2-Gram: <br>
1. She was running. and He was running.  &emsp;&ensp;    Score: 0.9333333333
2. They were walking. and They were talking. Score: 0.8333333333333334
3. We were playing. and You were playing.   &ensp;&nbsp; Score: 0.8125
4. We were cheering. and You were cheering.  Score: 0.8125
5. He was sleeping. and I was sleeping.      &emsp;&emsp;&ensp;Score: 0.8125

3-Gram: <br>
1. She was running. and He was running. Score: 0.9285714285714286
2. He was sleeping. and I was sleeping. Score: 0.8 
3. I was watching. and He was watching. Score: 0.7857142857142857
4. I was digging. and He was digging. Score: 0.7857142857142857
5. I was singing. and He was singing. Score: 0.7692307692307693

word-Gram: <br>
1. The birds were singing. and The birds were chirping. Score: 0.5 
2. The clouds were dark. and The clouds were grey. Score: 0.5 
3. The sun was setting. and The sun was shining. Score: 0.5 
4. I was building. and I was dreaming. Score: 0.3333333333333333
5. I was building. and I was singing. Score: 0.3333333333333333

#### Short Sentences
2-Gram: <br>
1. The cricket chirped in the night. and The crickets chirped in the night. Score 0.9 
2. The leaves rustled in the breeze. and The leaves rustled in the gentle breeze. Score: 0.875 
3. The clouds drifted across the sky. and The clouds drifted slowly across the sky. Score: 0.8529411764705882 
4. The stars twinkled in the night sky. and The stars twinkled and glittered in the night sky. Score: 0.725 
5. The horse galloped across the meadow. and The horses galloped in the meadow. Score: 0.7222222222222222

3-Gram: <br>
1. The cricket chirped in the night. and The crickets chirped in the night. Score: 0.84375 
2. The leaves rustled in the breeze. and The leaves rustled in the gentle breeze. Score: 0.8285714285714286 
3. The clouds drifted across the sky. and The clouds drifted slowly across the sky. Score: 0.7631578947368421
4. The stars twinkled in the night sky. and The stars twinkled and glittered in the night sky. Score: 0.7111111111111111 
5. The stars twinkled in the dark sky. and The stars twinkled in the night sky. Score: 0.6578947368421053

Word-Gram: <br>
1. The cricket chirped in the night. and The cricket chirped in the grass. Score: 0.6666666666666666 
2. The owl hooted in the distance.and The owl hooted in the tree. Score: 0.6666666666666666
3. The clouds drifted across the sky. and The clouds drifted slowly across the sky. Score: 0.5714285714285714
4. The leaves rustled in the breeze. and The leaves rustled in the gentle breeze. Score: 0.5714285714285714
5. The stars twinkled in the night sky. and The stars twinkled and glittered in the night sky. Score: 0.5555555555555556

#### Long Sentences
2-Gram <br>
1. Taking a deep breath, she smiled and embraced the journey ahead, confident that no matter what happened, she was strong enough to make it through. and Taking a deep breath, she opened her eyes and looked out across the fields, knowing that no matter what the future held, she was strong enough to face it. Score: 0.5897435897435898

2. The sun shone brightly on the dewy meadow, and a gentle breeze blew through the trees, bringing with it the scent of fresh grass and wildflowers as it stirred the leaves and rustled the long green blades. and The sun shone brightly through the window, and the birds sang in the trees, bringing a feeling of peace and contentment to the room as the gentle breeze rustled the curtains. <br>Score: 0.5725806451612904

3. The stars twinkled in the night sky, and a crescent moon shone its silvery light down onto the meadow, illuminating the dark shadows of the trees and the grass beneath them. and The stars twinkled in the night sky, and the moon shone down, casting its silver light on the lake that stretched out before her, its ripples sparkling in the moonlight.  <br>Score: 0.5403225806451613

4. She had been through so much in her life and was finally realizing that no matter what she had gone through, she would always have the chance to make her life better.and She knew that no matter what happened, she was strong enough to get through it, and she was ready to face whatever the world had in store for her. Score: 0.5277777777777778

5. Taking a deep breath, she smiled and embraced the journey ahead, confident that no matter what happened, she was strong enough to make it through. and She knew that no matter what happened, she was strong enough to get through it, and she was ready to face whatever the world had in store for her. Score: 0.5178571428571429

3-Gram: <br>
1. Taking a deep breath, she smiled and embraced the journey ahead, confident that no matter what happened, she was strong enough to make it through. and Taking a deep breath, she opened her eyes and looked out across the fields, knowing that no matter what the future held, she was strong enough to face it. Score: 0.3978494623655914

2. He took a deep breath, the smell of the freshly blooming flowers in the air, and the sound of the nearby stream flowing in the distance. and She walked through the city streets, the sound of the cars honking in the distance, and the smell of the freshly baked bread wafting through the air. Score: 0.3790849673202614

3. He stood in the middle of the forest, the sound of the birds singing in the trees, and the smell of the pine needles in the air. and She stood in the middle of the field, the smell of the freshly cut hay in the air, and the sound of the crickets chirping in the grass. Score: 0.3787878787878788

4. He took a deep breath, the smell of the freshly blooming flowers in the air, and the sound of the nearby stream flowing in the distance. and He took a deep breath, the crisp autumn air filling his lungs, and the smell of the freshly fallen leaves in the air. <br>Score: 0.3708609271523179

5. The stars twinkled in the night sky, and a crescent moon shone its silvery light down onto the meadow, illuminating the dark shadows of the trees and the grass beneath them. and The stars twinkled in the night sky, and the moon shone down, casting its silver light on the lake that stretched out before her, its ripples sparkling in the moonlight. Score: 0.3681592039800995

Word-Gram: <br>
1. He took a deep breath, the smell of the freshly blooming flowers in the air, and the sound of the nearby stream flowing in the distance. and He took a deep breath, the crisp autumn air filling his lungs, and the smell of the freshly fallen leaves in the air. <br>Score: 0.3235294117647059 

2. He stood in the middle of the forest, the sound of the birds singing in the trees, and the smell of the pine needles in the air. and She stood in the middle of the field, the smell of the freshly cut hay in the air, and the sound of the crickets chirping in the grass. Score: 0.2857142857142857

3. He took a deep breath, the smell of the freshly blooming flowers in the air, and the sound of the nearby stream flowing in the distance. and She stood in the middle of the field, the smell of the freshly cut hay in the air, and the sound of the crickets chirping in the grass. Score: 0.2777777777777778

4. Taking a deep breath, she smiled and embraced the journey ahead, confident that no matter what happened, she was strong enough to make it through. and Taking a deep breath, she opened her eyes and looked out across the fields, knowing that no matter what the future held, she was strong enough to face it. Score: 0.2682926829268293

5. The scent of freshly cut grass filled the air as the sun shone down on the meadow, the beauty of nature overwhelming in its simplicity. and The sun shone brightly through the window as she walked down the street, the birds chirped, the wind whistled, and the scent of freshly cut grass filled the air as she continued on her journey. Score: 0.2391304347826087

#### Spanish Sentences
2-Gram:
1. La tecnología también ha ayudado a reducir los costos de producción al permitir que las empresas realicen muchas tareas de forma más eficiente. and La tecnología también ha mejorado la seguridad al permitir que las empresas rastreen y controlen sus datos de forma más eficiente. Score: 0.6696428571428571 

2. El viento soplaba suavemente mientras él cerraba los ojos para disfrutar de la tranquilidad. and La brisa suave del mar acariciaba su rostro mientras él cerraba los ojos para disfrutar del momento. Score: 0.5764705882352941 

3. La tecnología también ha permitido a muchas empresas llevar a cabo transacciones de forma más segura con el uso de la criptografía. and La tecnología también ha mejorado la seguridad al permitir que las empresas rastreen y controlen sus datos de forma más eficiente. Score: 0.5655737704918032 

4. La tecnología ha cambiado la forma en que nos conectamos con los demás al ofrecer nuevas formas de comunicación, como el chat en vivo. and La tecnología ha cambiado la forma en que interactuamos con el mundo y ha proporcionado nuevas formas de hacer negocios. Score: 0.5566037735849056

5. La tecnología ha hecho que el mundo sea un lugar más seguro y conectado, lo que significa que ahora las personas pueden compartir información y conocimiento de forma más rápida y efectiva. and Los avances en la tecnología han creado nuevas formas de comunicación, como las redes sociales, que permiten a las personas conectarse en cualquier momento y lugar. <br>Score: 0.5460992907801419

3-Gram:
1. La tecnología también ha ayudado a reducir los costos de producción al permitir que las empresas realicen muchas tareas de forma más eficiente. and La tecnología también ha mejorado la seguridad al permitir que las empresas rastreen y controlen sus datos de forma más eficiente. Score: 0.47023809523809523

2. El viento soplaba suavemente mientras él cerraba los ojos para disfrutar de la tranquilidad. and La brisa suave del mar acariciaba su rostro mientras él cerraba los ojos para disfrutar del momento. Score: 0.452991452991453

3. El que no se arriesga, no pasa la mar. and El que no arriesga, no gana. Score: 0.45

4. La tecnología ha cambiado la forma en que nos conectamos con los demás al ofrecer nuevas formas de comunicación, como el chat en vivo. and La tecnología ha cambiado la forma en que interactuamos con el mundo y ha proporcionado nuevas formas de hacer negocios. Score: 0.40384615384615385

5. Todos los presentes se unieron a la fiesta, cantando, bailando y disfrutando de la compañía de sus vecinos. and El día de la fiesta, los vecinos se reunieron para disfrutar de la música y de la comida. Score: 0.37398373983739835

Word-Gram: <br>
1. La tecnología también ha ayudado a reducir los costos de producción al permitir que las empresas realicen muchas tareas de forma más eficiente. and La tecnología también ha mejorado la seguridad al permitir que las empresas rastreen y controlen sus datos de forma más eficiente. Score: 0.3125 

2. El que no se arriesga, no pasa la mar. and El que no arriesga, no gana. Score: 0.3

3. La tecnología ha cambiado la forma en que nos conectamos con los demás al ofrecer nuevas formas de comunicación, como el chat en vivo. and La tecnología ha cambiado la forma en que interactuamos con el mundo y ha proporcionado nuevas formas de hacer negocios. Score: 0.2727272727272727

4. El viento soplaba suavemente mientras él cerraba los ojos para disfrutar de la tranquilidad. and La brisa suave del mar acariciaba su rostro mientras él cerraba los ojos para disfrutar del momento. Score: 0.2608695652173913

5. El aire era fresco y limpio y la brisa suave del mar acariciaba su cara.and La brisa suave del mar acariciaba su rostro mientras él cerraba los ojos para disfrutar del momento. Score: 0.25

### Computational Cost
This experiment does not seem to have a heavy computational cost. everytime I ran any of the three k-grams code it never took longer then 6-7 seconds to complete. This tells me that it is not costing much to run these programs.